File: mobile/lib/features/audit/models/audit_log.dart
class AuditLog {
  final String id;
  final String actionType;
  final String module;
  final String recordType;
  final String recordId;
  final Map<String, dynamic>? oldData;
  final Map<String, dynamic>? newData;
  final String performedById;
  final String performedByName;
  final String userRole;
  final String shopId;
  final String shopName;
  final DateTime? timestamp;
  final String ipAddress;
  final String platform;
  final String remarks;
  final String status;

  const AuditLog({
    required this.id,
    required this.actionType,
    required this.module,
    this.recordType = '',
    this.recordId = '',
    this.oldData,
    this.newData,
    this.performedById = '',
    this.performedByName = '',
    this.userRole = '',
    this.shopId = '',
    this.shopName = '',
    this.timestamp,
    this.ipAddress = '',
    this.platform = '',
    this.remarks = '',
    this.status = 'success',
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    final performedBy = json['performedBy'];
    String pId = '';
    String pName = '';
    String pRole = '';
    if (performedBy is Map<String, dynamic>) {
      pId = (performedBy['id'] ?? performedBy['_id']).toString();
      pName = performedBy['name']?.toString() ?? '';
      pRole = performedBy['role']?.toString() ?? '';
    }
    final shop = json['shopId'];
    String sName = '';
    if (shop is Map<String, dynamic>) {
      sName = shop['name']?.toString() ?? '';
    }
    final device = json['deviceInfo'];
    String platform = '';
    if (device is Map<String, dynamic>) {
      platform = device['platform']?.toString() ?? '';
    }
    return AuditLog(
      id: (json['id'] ?? json['_id']).toString(),
      actionType: json['actionType']?.toString() ?? '',
      module: json['module']?.toString() ?? '',
      recordType: json['recordType']?.toString() ?? '',
      recordId: json['recordId']?.toString() ?? '',
      oldData: (json['oldData'] is Map<String, dynamic>) ? (json['oldData'] as Map<String, dynamic>) : null,
      newData: (json['newData'] is Map<String, dynamic>) ? (json['newData'] as Map<String, dynamic>) : null,
      performedById: pId,
      performedByName: json['performedByName']?.toString() ?? pName,
      userRole: json['userRole']?.toString() ?? pRole,
      shopId: json['shopId'] is String ? json['shopId'] as String : ((json['shopId'] as Map<String, dynamic>?)?['id'] ?? '').toString(),
      shopName: json['shopName']?.toString() ?? sName,
      timestamp: DateTime.tryParse(json['timestamp']?.toString() ?? ''),
      ipAddress: json['ipAddress']?.toString() ?? '',
      platform: platform,
      remarks: json['remarks']?.toString() ?? '',
      status: json['status']?.toString() ?? 'success',
    );
  }
}

File: backend/src/models/AuditLog.js
const mongoose = require('mongoose');

const ACTION_TYPES = [
  'CREATE_PRODUCT',
  'UPDATE_PRODUCT',
  'DELETE_PRODUCT',
  'RESTORE_PRODUCT',
  'CREATE_CATEGORY',
  'UPDATE_CATEGORY',
  'DELETE_CATEGORY',
  'RESTORE_CATEGORY',
  'STOCK_IN',
  'STOCK_OUT',
  'STOCK_TRANSFER',
  'CREATE_SALE',
  'UPDATE_SALE',
  'DELETE_SALE',
  'RESTORE_SALE',
  'CREATE_EXPENSE',
  'UPDATE_EXPENSE',
  'DELETE_EXPENSE',
  'RESTORE_EXPENSE',
  'CREATE_USER',
  'UPDATE_USER',
  'LOGIN',
  'LOGOUT',
  'REPORT_DOWNLOAD',
];

const MODULES = [
  'auth',
  'products',
  'categories',
  'inventory',
  'sales',
  'expenses',
  'shops',
  'users',
  'reports',
  'system',
];

const auditLogSchema = new mongoose.Schema(
  {
    actionType: { type: String, enum: ACTION_TYPES, required: true },
    module: { type: String, enum: MODULES, required: true },
    recordType: { type: String, default: '' },
    recordId: { type: mongoose.Schema.Types.Mixed, default: null },
    oldData: { type: mongoose.Schema.Types.Mixed, default: null },
    newData: { type: mongoose.Schema.Types.Mixed, default: null },
    performedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    performedByName: { type: String, default: '' },
    userRole: { type: String, enum: ['admin', 'manager', 'system'], default: 'manager' },
    shopId: { type: mongoose.Schema.Types.ObjectId, ref: 'Shop', default: null },
    shopName: { type: String, default: '' },
    timestamp: { type: Date, default: Date.now },
    ipAddress: { type: String, default: 'unknown' },
    deviceInfo: {
      platform: { type: String, default: '' },
      userAgent: { type: String, default: '' },
      appVersion: { type: String, default: '' },
    },
    remarks: { type: String, default: '' },
    status: {
      type: String,
      enum: ['success', 'deleted', 'restored'],
      default: 'success',
    },
  },
  { timestamps: true }
);

// Audit logs are immutable - only admin may read. Prevents modification/deletion.
auditLogSchema.pre('updateOne', function (next) {
  const error = new Error('Audit logs are immutable and cannot be modified.');
  error.name = 'ImmutableDocumentError';
  next(error);
});
auditLogSchema.pre('updateMany', function (next) {
  const error = new Error('Audit logs are immutable and cannot be modified.');
  error.name = 'ImmutableDocumentError';
  next(error);
});

auditLogSchema.index({ timestamp: -1 });
auditLogSchema.index({ performedBy: 1, timestamp: -1 });
auditLogSchema.index({ shopId: 1, timestamp: -1 });
auditLogSchema.index({ actionType: 1 });

module.exports = mongoose.model('AuditLog', auditLogSchema);
module.exports.ACTION_TYPES = ACTION_TYPES;
module.exports.MODULES = MODULES;
